File: examples/C/openmp_converted/blur_simple_solution_omp.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
/**
 * Apply a gaussian blurring filter to a picture generated on the fly
 *
 * List of functions:
 *   - void blur(unsigned char* pic,  unsigned char* blurred, size_t rows, size_t cols)
 *     the actual filter
 *   - void fill(unsigned char* pic, size_t rows, size_t cols)
 *     generate the original picture
 *   - void out_pic(unsigned char* pic, char* name, size_t rows, size_t cols)
 *     create a .rgb file 
 */

void blur(unsigned char* pic,  unsigned char* blurred, size_t rows, size_t cols)
{
    /**
     * Perform the blurring of the picture
     * @ param pic(in): a pointer to the original picture
     * @ param blurred(out): a pointer to the blurred picture
     */
   size_t i, j, l, i_c, j_c;
   unsigned int pix;
   unsigned char coefs[5][5] = { {1,  4,  6,  4,  1},
                                 {4, 16, 24, 16,  4},
                                 {6, 24, 36, 24,  6},
                                 {4, 16, 24, 16,  4},
                                 {1,  4,  6,  4,  1}};

#pragma acc parallel loop copyin(coefs[:5][:5],pic[:3*rows*cols]) copyout(blurred[:3*rows*cols])
#pragma omp target teams loop map(to:coefs[:5][:5],pic[:3*rows*cols])\
            map(from:blurred[:3*rows*cols])
   for (i=2; i<rows-2; ++i)
      for (j=2; j<cols-2; ++j)
         for (l=0; l<3; ++l)
         {
            pix = 0;
            for (i_c=0; i_c<5; ++i_c)
                for (j_c=0; j_c<5; ++j_c)
                   pix += (pic[(i+i_c-2)*3*cols+(j+j_c-2)*3+l]
                           *coefs[i_c][j_c]);

             blurred[i*3*cols+j*3+l] = (unsigned char)(pix/256);
         }
}
void fill(unsigned char* pic, size_t rows, size_t cols)
{
    /**
     * Fill the picture with data
     * @param pic(out): a pointer to the pixel to be blurred
     * @param rows(in) the number of rows in the picture
     * @param cols(in) the number of columns in the picture
     *
     */

    size_t i, j;
#pragma acc parallel loop copyout(pic[0:3*rows*cols])
#pragma omp target teams loop map(from:pic[0:3*rows*cols])
   for (i=0; i < rows; ++i)
      for (j=0; j < 3*cols; ++j)
         pic[i*3*cols+j] = (unsigned char) (i+(j%3)*j+i%256)%256;
    
    
}

void out_pic(unsigned char* pic, char* name, size_t rows, size_t cols)
{
    /**
     * Output of the picture into a sequence of pixel
     * Use show_rgb(filepath, rows, cols) to display
     * @param rows(in) the number of rows in the picture
     * @param cols(in) the number of columns in the picture
     */
   FILE* f = fopen(name, "wb");
   fwrite(pic, sizeof(unsigned char), rows*3*cols, f);
   fclose(f);
}

int main(void)
{
   size_t rows,cols;

   rows = 4000;
   cols = 4000;

   printf("Size of picture is %d x %d\n", rows, cols); 
   unsigned char* pic = (unsigned char*) malloc(rows*3*cols*sizeof(unsigned char));
   unsigned char* blurred_pic = (unsigned char*) malloc(rows*3*cols*sizeof(unsigned char));

   // Create the original picture
   fill(pic, rows, cols);

   // Apply the blurring filter
   blur(pic, blurred_pic, rows, cols);

   out_pic(pic, "pic.rgb", rows, cols);
   out_pic(blurred_pic, "blurred.rgb", rows, cols);

   free(pic);
   free(blurred_pic);

   return 0;
}

// Code was translated using: /home/very/bin/acc2mp -async=nowait -specify-language=C -generate-report blur_simple_solution.c

File: examples/Fortran/openmp_converted/blur_simple_solution_omp.f90
MODULE pixels
    USE OPENACC
use omp_lib
    IMPLICIT NONE
    CONTAINS
        SUBROUTINE fill(pic,rows,cols)
            INTEGER(kind=1),DIMENSION(0:),INTENT(OUT) :: pic
            INTEGER,INTENT(IN)                        :: rows,cols
            INTEGER                                   :: i,j, my_unit, val
            
            DO i=0, rows-1
                DO j=0, 3*cols-1
                     val = i+(MOD(j,3))*j+MOD(i,256)
                     pic(i*3*cols+j) = MOD(val,128)
                END DO
            END DO
        END SUBROUTINE fill

        SUBROUTINE blur(pic, blurred, rows, cols)
            INTEGER(kind=1),DIMENSION(0:),INTENT(IN)  :: pic
            INTEGER(kind=1),DIMENSION(0:),INTENT(OUT) :: blurred
            INTEGER,DIMENSION(0:4,0:4)               :: coefs
            INTEGER,INTENT(IN)                        :: rows,cols
            INTEGER(kind=4)                         :: i,j,i_c,j_c,l,pix
            INTEGER :: my_unit
            coefs(0,:)= (/ 1, 2, 3, 3, 1 /)
            coefs(1,:)= (/ 2, 8, 12, 8, 2 /)
            coefs(2,:)= (/ 3, 12, 14, 12, 3 /)
            coefs(3,:)= (/ 2, 8, 12, 8, 2 /)
            coefs(4,:)= (/ 1, 2, 3, 2, 1 /)
            !$acc parallel loop copyin(pic(0:), coefs(0:,0:)) copyout(blurred(0:))
!$omp target teams loop map(to:pic(0:),coefs(0:,0:))&
!$omp map(from:blurred(0:))
            DO i=2,rows-3
                DO j=2,cols-3
                    DO l=0,2
                        pix = 0
                        DO i_c=0,4
                            DO j_c=0,4
                                pix = pix + pic((i+i_c-2)*3*cols+(j+j_c-2)*3+l)*(coefs(i_c,j_c)) 
                            END DO
                        END DO
                        blurred(i*3*cols+j*3+l)=pix/128
                    END DO
                END DO
            END DO
        END SUBROUTINE blur

        SUBROUTINE out_pic(pic, name)
            INTEGER(kind=1),DIMENSION(0:),INTENT(IN) :: pic
            CHARACTER(len=*),INTENT(IN)              :: name  
            INTEGER                                  :: my_unit

            OPEN(NEWUNIT=my_unit, FILE=name, status='replace' ,ACCESS="stream", FORM="unformatted")
            WRITE(my_unit,POS=1) pic 
            CLOSE(my_unit)
        END SUBROUTINE out_pic

end module pixels

PROGRAM blur_pix
    use PIXELS
    implicit none

    integer                                      :: rows, cols, i, long, j, check, numarg
    integer  (kind=1), dimension(:), allocatable :: pic,blurred_pic
    character(len=: ), allocatable               :: arg
    integer, dimension(2)                        :: n
 
    rows = 4000
    cols = 4000

    write(6,"(a23,i7,a2,i7)") "Size of the picture is ",rows," x",cols
    allocate(pic(0:rows*3*cols), blurred_pic(0:rows*3*cols))

    call fill(pic,rows,cols)
    call blur(pic, blurred_pic, rows, cols)

    call out_pic(pic, "pic.rgb")
    call out_pic(blurred_pic, "blurred.rgb")

    deallocate(pic, blurred_pic)
END PROGRAM blur_pix

! Code was translated using: intel-application-migration-tool-for-openacc-to-openmp blur_simple_solution.f90
